implement purescript-jtable

brainrape opened the pull request:
Here's my implementation of purescript-jtable.
It has all the high-level features in the requirements. Some things are implemented differently though, and I hope you will accept my reasoning below.

I recommend to [Try it](http://rawgit.com/brainrape/purescript-jtable/dev-squash/examples/try.html) with different renderers and  example data (including the "Comprehensive" example) right away, look at [example usage](https://github.com/brainrape/purescript-jtable/blob/dev-squash/src/Data.Json.JTable.Examples.purs) and then read the analysis below.

### Types, header

The label hierarchy is unified based on its shape. It becomes a hierarchy of nodes with width. Primitives have width 1, tuples more. The concrete types of the primitives are irrelevant, only the width is used. Arrays are not represented, they are unified into the most general shape hierarchy. Thus it is ensured that all data present in the Json can be rendered sensibly.

Contrast this to the spec, which says "Classify every leaf node using JSemantic [...], based on the most common type of semantic detected in the data". The problems with this step include that a statistical method is suggested to determine the type of data, with no hints as to how to handle divergent leaves. Also, leaf node semantic type, and even distinguishing between primitive json types is irrelevant to the header generation, where only tuple width and label hierarchy matters. In fact, JSemantic is a different issue, independent of JTable (but which combines nicely), as discussed below.

### Tuples

There were some undefined aspects of handling tuples. Here's how I chose to resolve them:

Homogenous arrays of equal sizes are rendered vertically, except when their size equals 2, in which case they are rendered horizontally. This is somewhat arbitrary, but usefult for geographic coordinates.

Heterogenous arrays of different sizes are rendered horizontally, with the maximum tuple size becoming the column width

Lists of objects with dijunct keys will be displayed in one row (treating them as tuples of objects).


### TableStyle.th 

Another difference is: JPath is an array of strings. Arrays (which could have different nestings) are not represented in the header hierarchy, so a [String] suffices for the path instead of the unwieldier JCursor.

```purescript
type JPath = [String]
```

The header rendering function gets a JPath as an argument.

```purescript
TableStyle.th :: JPath -> Markup
```

The default implementation is equivalent to
```purescript
noStyle.th = \p -> th $ text $ Data.Array.Unsafe.last p
```

Where `th` and `text` are from Smolder and the path is guaranteed to have at least one element.

### Column Ordering

The `ColumnOrdering` type has been simplified, `inOrdering` is a value. Easy to roll your own.

```
type ColumnOrdering = JPath -> JPath -> Ordering 
inOrdering    :: ColumnOrdering
alphaOrdering :: ColumnOrdering
```

### TableStyle.td

JTable does not depend on JSemantic. In fact, I recommend releasing JSemantic as a separate library, since it is orthogonal to JTable and separate from Argonaut (although it uses the JsonPrim type). However, for now I have inluded it here.

Accordingly, `td` receives a JsonPrim argument that is easy to use.

```purescript
TableStyle.td :: JCursor -> JsonPrim -> Markup
```

### JSemantic

Data constructors of the type JSemantic have parameters and carry their data with them, converted to a useful type.

```purescript
data JSemantic = Integral   Number
               | Fractional Number
               | Date       Date.Date
               | DateTime   Date.Date
               | Time       Date.Date
               | Interval   Date.Date Date.Date
               | Text       String
               | Bool       Boolean
               | Percent    Number
               | Currency   Number
               | NA
```
Although String might be a better choice for Currency and Percent, to avoid loss of precision.

And it's easy to combine JTable and JSemantic:

```purescript
exampleRendererSemantic = renderJTable defJTableOpts {
    style = noStyle { td = \c j -> case toSemantic j of
      Currency n -> td ! className "money" $ text $ show n
  }}
```

Rendering differently based on path is also easy, dispatching on the JCursor argument, using operations from Argonaut.


### Also,

There are unit tests for some edge cases that I fixed along the way, and StrongCheck tests are run as well.

The code is not exceptionally fast or clean and simple, but it's not slow or too complex either. It gets the job done.

I'm sure the examples, the [readme](https://github.com/brainrape/purescript-jtable/blob/dev-squash/README.md), and the above will convince you that the implementation provides a straightforward, intuitive, and correct conceptual mapping between any Json schema and the resulting table, has features equivalent to the requirements, with the differences backed by sound analysis.

Let me know if changes or additions are needed.

jdegoes commented:
@brainrape This looks really great! It's probably going to take me another day to work all the way through this (and add some minor comments), but after playing around with the demo, it looks like it's well thought out and will suit our needs. @cryogenian has made some comments above, I should have all mine in by tomorrow -- all fairly minor stuff.

jdegoes commented:
OK, about ready to dig into this. One thing I noticed when testing the application is that it only horizontally renders 2-tuples, not arbitrary n-tuples (so, for example, if I have a homogeneous schema and a given field is a 3D point, all the coordinates will render vertically -- somewhat arbitrarily). This makes some JSON render awkwardly and take up a lot of space. I'll comment on that again wherever that code is, but this is something we'll want to address.

jdegoes commented on examples/Data.Json.JTable.Examples.purs:
```diff
@@ -0,0 +1,98 @@
+
+
+foreign import _main """
+function _main (render_markup) { return function (examples) {
```
Any reason why this function can't be written in (mostly or entirely) Purescript? The logic is doubly hard to follow since it's Javascript embedded in a string. Think most (all?) of this FFI functionality is in `simple-dom` or `timers`.

brainrape replied:
I moved most of this to Purescript. I didn't use `purescript-simple-dom` but FFI'd a few DOM functions.

jdegoes commented on examples/try.html:
```diff
@@ -0,0 +1,1471 @@
+    "replyTo":  [9654726, "A44124F"],
+    "time":     "2015-03-01"
+  }]
+}
```
These are great! It would also be good to have a 3 tuple example, and maybe [this real-world example](https://dev.twitter.com/rest/reference/get/search/tweets).

brainrape replied:
As explained in my first comment, 
- heterogeneous arrays (tuples) are always rendered horizontally
- homogeneous arrays of length 2 are rendered horizontally
- homogeneous arrays of length >2 are rendered vertically

Unification still applies. If both are encountered, the arrays will be rendered vertically, as this is the most general.

It's probably a good idea to turn the maximum width of homogeneous tuples into a config option.

jdegoes commented on package.json:
```diff
@@ -1,15 +1,17 @@
-  "name":       "purescript-jtable",
-  "license":    "MIT",
+  "name": "purescript-jtable",
+  "version": "0.0.1",
```
You may want to bump this to 0.1.0 for better compatibility with semantic versioning (otherwise, they'll be no way to do backward-compatible changes).

brainrape replied:
Ok, I'll bump the version after some fixes.

jdegoes commented on src/Data.Json.JSemantic.purs:
```diff
@@ -0,0 +1,105 @@
+import Data.String
+import Data.Foldable
+import qualified Data.Array as A
+import qualified Data.Array.Unsafe as AU
```
I'll comment on this below but let's try to see if we can eliminate use of this (it's usually possible and sometimes cleans up the code).

brainrape replied:
fixed

jdegoes commented on src/Data.Json.JSemantic.purs:
```diff
@@ -0,0 +1,105 @@
+import Math
+
+
+data JSemantic = Integral   Number
```
This looks good and means we don't have to pass Json to the styler, which is nice. :+1:

jdegoes commented on src/Data.Json.JSemantic.purs:
```diff
@@ -0,0 +1,105 @@
+               | Date       Date.Date
+               | DateTime   Date.Date
+               | Time       Date.Date
+               | Interval   Date.Date Date.Date
```
I'm not sure we can represent an interval using two dates. We really need a good `purescript-duration` library, but no such thing exists (although there's `purescript-moment`, and I know the underlying JS library [has support for durations](http://momentjs.com/docs/#/durations/)). ISO8601 defines the string representation for them. 

[There are some terrifying](https://gist.github.com/philipashlock/8830168) regexes to parse duration, but not sure they'd be that useful, because the renderer would need a typed ADT in order to provide a useful styling of things like "every other week".

Not sure if we want to address this in this PR unless there's a quickfix, otherwise I'm fine ignoring intervals for now. Can always add them later after we have a good PS duration library.

brainrape replied:
That might be the case for durations, but an Interval (as suggested in the spec) is "part of the time axes limited by two instants" ([wp](http://en.wikipedia.org/wiki/Time#Time-like_concepts:_terminology)), and two dates are an adequate representation.

We should choose a `(date|time|datetime)Regex` that is in line with ISO8601 but is sufficiently strict (eg. won't parse not fully specified dates like simple numbers). 

We can also make regexes configurable, and split JSemantic into a separate library. Are you ok with that?

jdegoes commented on src/Data.Json.JSemantic.purs:
```diff
@@ -0,0 +1,105 @@
+               | Currency   Number
+               | NA
+
+instance showJSemantic :: Show JSemantic where
```
It's general convention that a `Show` instance should render a string that is code which can reconstruct value being shown (e.g. `Show (Integral n)` would be defined as `"Integral " ++ show n`). For other uses, it's best to just define a custom function (e.g. `renderJSemantic`) which has whatever string rendering you like.

brainrape replied:
fixed, there's `renderJSemantic` now, and `showJSemantic` renders code

jdegoes commented on src/Data.Json.JSemantic.purs:
```diff
@@ -0,0 +1,105 @@
+percentRegex = test $ regex """^-?\d+(\.\d+)?\%$""" noFlags
+
+currencyRegex :: String -> Boolean
+currencyRegex = test $ regex "^\\$?([0-9]{1,3},([0-9]{3},)*[0-9]{3}|[0-9]+)(.[0-9][0-9])?$" noFlags
```
Can you make this one a triple string too, like the others?

jdegoes replied:
Also, see below note on the `$` symbol and other currency symbols.

brainrape replied:
fixed

jdegoes commented on src/Data.Json.JSemantic.purs:
```diff
@@ -0,0 +1,105 @@
+foreign import parsePercent """function parsePercent (s) {
+  return s.replace("%", "") * 1}""" :: String -> Number
+
+foreign import parseCurrency """function parseCurrency (s) {
```
Aside from `$`, you can catch all unicode currency symbols by looking for characters in the range 20A0–20CF.

Beyond that, this function seems fishy, because its type claims to be able to convert any string at all into a number (by the type I'd guess it computes the string length, which is the only way I can think of to reasonably satisfy that promise).

brainrape replied:
Technically, NaN is a Number :)
I'll probably upgrade these to maybes.

brainrape replied:
fixed, there are additional currency symbols in the 'Symbol, Currency' category.
http://www.fileformat.info/info/unicode/category/Sc/list.htm

brainrape replied:
fixed, there are additional currency symbols:
http://www.fileformat.info/info/unicode/category/Sc/list.htm

jdegoes commented on src/Data.Json.JSemantic.purs:
```diff
@@ -0,0 +1,105 @@
+    Nothing -> Text s -- shouldn't happen
+  else 
+    let sp = split " - " s in 
+    if A.length sp == 2 && all dateRegex sp 
```
The proper way to do this, which will eliminate the unsafe `head`, is to use pattern matching, e.g.:

```purescript
case split " - " in of
  start : end : [] -> Interval <$> Date.fromString start <*> Date.fromString end
  _ -> Nothing
```

This gives you a `Maybe Interval` which you can then use `fromMaybe` on to retrieve either the `Interval` (if it's `Just`), or a default value (`Text s`).

brainrape replied:
fixed

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+import Data.String (joinWith, localeCompare)
+import Data.Array
+import Data.Foldable
+import qualified Data.Array.Unsafe as AU
```
Ditto on this, let's see if we can eliminate this and cleanup the code at the same time!

brainrape replied:
I still use `Array.Unsafe.head` for `JPath`, otherwise fixed.

brainrape replied:
fixed that as well

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+import Text.Smolder.Markup (Markup(..), Attribute(..), attribute, (!))
+
+-- path of object keys, with array indices omitted
+type JPath = [String]
```
I like this, makes it nice and simple to work with. Although if you have the guarantee that there's always a head (which I think I recall from the PR comments), then we can represent this as `Tuple String [String]`. 

If you want a "true" non-empty list type, then there's `Cofree Maybe` from the `purescript-free` package.

brainrape replied:
`Tuple` won't do since the root node has an empty path (but `head` only gets called on child nodes' paths).

brainrape replied:
I'll leave this for later.

brainrape replied:
fixed

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+type JPath = [String]
+
+-- rows of cells
+type Table = [[Cell]]
```
Do we have to put the whole JSON object into cells, or is there an incremental process we could use when building the markup, which would allow us to cut down on intermediate storage?

brainrape replied:
`cFromJson` recursively merges lists of tables. I guess if we rewrite that to work in terms of Markup, it could work, but I'm not sure we'd gain much in terms of storage.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+
+-- header data
+data Tree = T JPath Number Number [Tree]
+tPath (T p _ _ _) = p
```
This is totally fine for this PR, although you may be interested in checking out purescript-lens for future projects. They make short work of getting and "setting" members of a data value.

jdegoes replied:
On further reflection, it occurs to me that the row and column numbers here are implicit -- we don't need to store them, since they can always be computed from the top-down. Removing the row and column numbers will simplify a lot of code and remove the possibility that they aren't set correctly.

jdegoes replied:
Also, instead of storing a complete `JPath`, remove potential for error by storing just a `String`, i.e. the label for this point in the node hierarchy. The full `JPath` can be computed trivially in a top-down fashion, just as for row/column number.

brainrape replied:
Those are not row and column numbers, but label width (colspan) and height (max levels of children). They are not implicit but calculated bottom-up.
JPath could be computed top-down just before rendering (eg `toRows`), but it was simpler to build up in one go, and the public api uses it as well.
All the above are computed in one traversal.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+tHeight (T _ _ h _) = h
+tKids (T _ _ _ k) = k
+
+instance showTree :: Show Tree where
```
Again, since this does not render code, it should be a helper function instead of a `Show` instance, to be consistent with how `Show` is used in Purescript.

brainrape replied:
This renders code.
I think this is fine to leave here, since the Tree type is not exported by `Data.Json.JTable`, only `Data.Json.JTable.Internal` (which is used for testing), and it never leaves the API. Otherwise a recursive render function would have to be defined that in turn couldn't make use of showArray.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+cJsonPrim (C _ _ _ j) = j
+
+instance showCell :: Show Cell where
+  show (C c w h j) = joinWith " " ["(C", show w, show h, show c, show j, ")"]
```
Ditto.

brainrape replied:
This also renders code.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+  show (C c w h j) = joinWith " " ["(C", show w, show h, show c, show j, ")"]
+
+
+foreign import jnull "var jnull = null;" :: Json
```
This already exists in Argonaut, I think it's called `jNull`.

brainrape replied:
We'd need `jnull = (fromNull (n::JNull))::Json` but no constructor is exported for JNull, one can only get a `JNull` from `foldJson` if one feeds it exactly a value like `jnull`.

brainrape replied:
`jnull = (primToJson primNull) :: Json` also works, i'll use that.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+foreign import jnull "var jnull = null;" :: Json
+
+_cN = const primNull
+toPrim = (foldJson _cN primBool primNum primStr _cN _cN) :: Json -> JsonPrim
```
This converts all non-primitives to null values. Are you sure that's a sensible thing to do? It could lead to misuse at a higher layer. Instead, it might be better to push this up, so that higher level code, which knows it isn't an object or array, can take advantage of that knowledge to only convert the primitives to a `JsonPrim`.

brainrape replied:
Updated, toPrim returns Maybe JsonPrim now.

brainrape replied:
`toPrim` now returns `Maybe JsonPrim`.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+toPrim = (foldJson _cN primBool primNum primStr _cN _cN) :: Json -> JsonPrim
+zipWithIndex xs = zip xs (0 .. length xs - 1) 
+
+so :: forall a. Boolean -> a -> Maybe a
```
I've got to break here for now, will resume later (probably tomorrow) from this point.

brainrape replied:
I renamed this function to justIf - it returns a `Just` if the boolean is true.

brainrape replied:
And removed it, using something else.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+-- maybe return the width of a tuple composed of primitive values
+widthOfPrimTuple :: JPath -> [Json] -> Maybe Number
+widthOfPrimTuple path ja = 
+  path !! 0 *> ja !! 1 *> let
```
1. If the tuple is composed of primitive values, then we should use `[JsonPrimitive]` here, instead of `Json`, to offer enhanced type safety.
2. Currently, this judges a thing to be a tuple if all the elements have the same type. This is sort of a "horizontal" definition of uniformity that would exclude many real-world tuples. Instead, we need to detect if there is "vertical" uniformity, i.e. all fields within the table have the same structure (e.g. the array field is always length 4 or null, and the types line up vertically).
3. As side notes, the `const` function can be used to simplify the `foldJson` application.

brainrape replied:
1. We don't know if that is the case. The function returns Nothing if there are non-primitives.
2. Can you give an example? Actually if all elements have the same type, it's an array (rendered vertically), and if elements have different types, then it's a tuple (rendered horizontally)
3. how? `foldJson (const 0) (const 1) (const 2) (const 3) (const 4) (const 4)` doesn't look much simpler.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+
+
+-- add child to tree, unify if exists
+tMergeArray :: Tree -> Tree -> Tree
```
This code can be greatly simplified if we remove row and column numbers from `Tree`. Then we only have to find the greatest common prefix, and there is no possibility the row/col logic can go awry (the row/col are trivial to compute in a top-down fashion).

brainrape replied:
These are cell widths/heights, not row and column numbers.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+               in T p w' h' k'
+
+-- produce a tree of header data from json
+tFromJson :: JPath -> Json -> Tree
```
The high-level structure of this function can be simplified by, instead of trying to convert the `Json` to each type (object, array, etc.), and then later dealing with the `Nothing` case, folding over the `Json` using `foldJson`, so each step produces exactly what it is supposed to.

jdegoes replied:
Also, this code will get quite simpler by removing the explicit row/col in `Tree`.

brainrape replied:
Updated to foldJson.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+             h = 1 + foldl max 0 (k <#> tHeight)
+         in T path w h k
+  m = (>>=) (toArray json)
+  tuple = m \a -> (widthOfPrimTuple path a) <#> \n -> T path n 0 []
```
This code will have to change to detect and manage "vertically type compatible" tuples.

brainrape replied:
Please give me an example.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+
+
+-- merge table segments for each key of an object into one
+cMergeObj :: [(Tuple Number Table)] -> Table
```
Storing explicit row / col is almost always a warning sign that the code is duplicating information content and can be simplified (which usually ends up preventing whole classes of "in sync"-type bugs).

brainrape replied:
The Number in the type here is the maximum width of each table (which we know from before, so it is passed to this function). Row/col numbers are implicit, cell height/width is not.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+cMergeObj rss = let
+  maxh = rss <#> snd <#> length # foldl max 0
+  in (0 .. maxh-1) <#> \n -> rss >>= uncurry \w rs ->
+    if length rs == 1 then if n == 0 
```
Instead of checking the length, use pattern matching -- this will also eliminate the unsafe `AU.head` below. Also, `rs !! n` and `fromMaybe` logic is duplicated, so you could allow the `Maybe` to survive a bit longer and factor out that duplication at the top level.

brainrape replied:
fixed

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+
+-- maybe merge a tuple of objects into a table segment
+mergeObjTuple ::Tree -> JCursor -> [Json] -> Maybe Table
+mergeObjTuple t@(T p w h k) c ja = head ja *> do
```
The head of `ja` is being checked but not used, at least, not used here. There's no reason to do this as the following code will produce `Nothing` if the array is empty (thanks to `for` working in the `Maybe` monad).

brainrape replied:
Actually, using `for` on an empty `[Maybe a]` produces `Just []`, I guess because it doesn't actually encounter any `Nothing`s. I fell for it too.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+  let all_keys = concat keyss
+  so ((length all_keys) == (length $ nub all_keys)) $ 
+    cMergeObj $ k <#> \(t'@(T p' w' _ _)) -> let 
+      label = AU.last p'
```
This is an example of a place where, if we change the representation of `JPath`, and adopt, `Tuple String [String]`, or use a real non-empty list like `Cofree Maybe String`, then we can eliminate this unsafe `AU.last`.

brainrape replied:
I'll leave this for later.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+
+-- produce data table from json, according to header tree
+cFromJson :: Tree -> JCursor -> Json -> Table
+cFromJson t@(T p w h k) c json = let
```
Again, I always try to avoid the `Maybe` monad when possible, because code so written is harder to reason about because any `Nothing` anywhere in a chain can cause the whole thing to be `Nothing`. It's sort of the Javascript of functional programming. In the case of this particular function, we can eliminate a lot of this by using `foldJson`, and combining the width and tuple computation. Also making `toPrim` type-accurate will have some effects on this function (cleaning it up, I believe, and making it easier to reason about).

brainrape replied:
updated to foldJson

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+
+-- render a grid from an array of arrays
+renderRows :: forall a. (Markup -> Markup) -> (Number -> Number -> a -> Markup) -> [[a]] -> Markup
+renderRows tr' cellf rows = mconcat $ do
```
This is a good example of keeping row/col implicit "until the last possible moment", which means this code can have no errors about row/col.

brainrape replied:
All rendering goes through this function.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+
+-- produce header rows from header tree
+tsToRows :: [Tree] -> [[Tree]]
+tsToRows ts = if null ts then [] else ts : tsToRows (ts >>= tKids)
```
Boolean checks are usually a sign the code can be written better, e.g.:

```purescript
tsToRows [] = []
tsToRows (x : xs) = x : tsToRows (xs >>= tKids)
```

Also note that information is being duplicated here, which will go away when `Tree` stops storing row/col information.

brainrape replied:
Fixing. I wanted to keep a one liner as such.
Only an array of array of references is built up, so it doesn't use much space.

brainrape replied:
It's actually
```
tsToRows [] = []
tsToRows ts@(_:_) = ts : tsToRows (ts >>= tKids)
```
I'm not sure it looks any better thatn `if null`.

brainrape replied:
As commented below, pattern matching doesn't have any benefit here since we're not using the inferred information statically. Also, the array gets copied just to be discarded immediately, after going through the same length check.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+
+renderThead :: (Markup -> Markup) -> (Tree -> Markup) -> Tree -> Markup
+renderThead tr' thf (T p w h k) =
+  let rs i k = if null k then h - i else 1
```
See above comments on booleans, this can be refactored to avoid the if statement and make the code cleaner.

brainrape replied:
We don't actually use values from k, and a pattern match would entail more code and translate to the same thing. I like `if null` better here, but i can fix it if it's an eyesore for you.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+  in renderRows tr' tdf' $ tsToRows k
+
+renderTbody :: (Markup -> Markup) -> (Cell -> Markup) -> Tree -> Json -> Markup
+renderTbody tr' tdf t json =
```
An option it would be nice to have is to separate the header table from the body table. Why? So you can apply fixed positioning to the header, so if you scroll, the header stays stuck to the top.

brainrape replied:
the tables are separate, in each of `thead` and `tbody` elements, so css can be applied. Although there is no separate render function for them in `TableStyle`, you can eg. set a class on `table` and use css selectors `.table > thead` and `.table > tbody` to style them.

jdegoes commented on src/Data.Json.JTable.Internal.purs:
```diff
@@ -0,0 +1,214 @@
+-- pad tall header cells from above
+insertHeaderCells :: Number -> Tree -> Tree
+insertHeaderCells maxh t@(T p w h k) = 
+  if null k 
```
Should be able to eliminate this if test.

brainrape replied:
Fixing.

brainrape replied:
The pattern matching version would look like this:
```
insertHeaderCells maxh t@(T p w h k) = case k of
  (_:_) -> T p w h (k <#> insertHeaderCells (maxh - 1))
  _     -> if maxh > 0 then T [] w 1 [insertHeaderCells (maxh - 1) t]
                       else T p w 1 k
```
Of which the pattern match translates to
```
if (_33.value3.length >= 1) {
  var _120 = _33.value3.slice(1); 
/* ... */}
```
This would not only have the same length check but also copy the array, only to discard the copy. Usin pattern matching here doesn't bring any benefit, because no inferred information is used, it's just trying to keep up List-like appearances, but in the end we're working with Arrays, not Lists.

jdegoes commented on src/Data.Json.JTable.purs:
```diff
@@ -0,0 +1,74 @@
+  columnOrdering :: ColumnOrdering,
+  insertHeaderCells :: Boolean }
+
+defJTableOpts  = {
```
This is very handy, we have a convention of calling these `xDefault`, so this one should be called, `jTableOptsDefault`.

brainrape replied:
fixed

jdegoes commented on src/Data.Json.JTable.purs:
```diff
@@ -0,0 +1,74 @@
+  th = (\t -> o.style.th (t # tPath)),
+  td = (\c -> o.style.td (c # cCursor) (c # cJsonPrim)) }}
+
+renderJTableArray :: JTableOpts -> [Json] -> Markup
```
No need to include this function since the user can always wrap on their side.

brainrape replied:
Glad you agree (this is what I brought up in a pull request). Removed.

jdegoes commented on test/Data.Json.JTable.StrongCheck.purs:
```diff
@@ -0,0 +1,112 @@
+genJsonObject = sized \size -> 
+  (vectorOf size $ resize (size-1) arbitrary) <#> SM.fromList >>> A.fromObject
+
+genJson :: Gen Json
```
This is nice. I think it may be better than the one in Argonaut because it uses the size parameter.

jdegoes commented on test/Data.Json.JTable.StrongCheck.purs:
```diff
@@ -0,0 +1,112 @@
+module Test.Data.Json.JTable.StrongCheck where
+
+import Data.Array
+import qualified Data.Array.Unsafe as AU
```
Let's try to eliminate this unsafeness.

brainrape replied:
fixed

jdegoes commented on test/Data.Json.JTable.StrongCheck.purs:
```diff
@@ -0,0 +1,112 @@
+
+-- repeat cells with [row|col]span 
+unprojectT :: forall a. (a -> Number) -> (a -> Number) -> [[a]] -> [[a]]
+unprojectT widthFn heightFn rows = if null rows then [] else let
```
See other comments, pattern matching is applicable here.

brainrape replied:
fixed

jdegoes commented on test/Data.Json.JTable.StrongCheck.purs:
```diff
@@ -0,0 +1,112 @@
+    row <- rows
+    return $ concat $ do 
+      cell <- row
+      return $ do
```
I think there are Array fill / repeat functions you can use in `Data.Array` -- you don't actually use the `n` below.

brainrape replied:
couldn't find it

jdegoes commented on test/Data.Json.JTable.StrongCheck.purs:
```diff
@@ -0,0 +1,112 @@
+  op = \rows nrow -> if null rows then [nrow] else snoc rows $ let
+    row = AU.last rows
+    op' = \(Tuple done rest) (Tuple cell h) ->
+      if h > 1 
```
This can deconstruct on `h` to eliminate the unsafe head & tail below, and clean up the code.

brainrape replied:
fixed

jdegoes commented on test/Data.Json.JTable.StrongCheck.purs:
```diff
@@ -0,0 +1,112 @@
+-- check that thead and tbody are the same width (or 0)
+lengthsOkT :: forall a b. [[a]] -> [[b]] -> Boolean
+lengthsOkT t1 t2 = 
+  (null t1) || (null t2) || (
```
See above comments on deconstruction to eliminate all these unsafe functions.

brainrape replied:
fixed

jdegoes commented on test/Data.Json.JTable.StrongCheck.purs:
```diff
@@ -0,0 +1,112 @@
+
+-- foreign import j0 """var j0 = null""" :: Json
+
+main = do
```
Since you already have a `Gen Json`, it could be useful to add a menu item to the test application: "random json", just to see how the renderer copes with some randomly generated json.

brainrape replied:
added

jdegoes commented on test/Data.Json.JTable.Unit.purs:
```diff
@@ -0,0 +1,97 @@
+import Data.Json.JTable
+
+
+foreign import jNull "var jNull = null" :: Json
```
This you can find in argonaut.

brainrape replied:
fixed

jdegoes commented:
The other thing is there is deep nesting in the module file names. It's a convention in Purescript to create folders, and only have nested 1-level. So, for example, `Data.Argonaut` would be located in `Argonaut.purs` file inside a `Data` directory. We should follow that convention here.

jdegoes commented:
OK, finally worked my way through this. Please submit follow-up commits and I'll review when I get a chance. Also reach out to me with any questions, if you want a quick answer, IRC is usually better (@jdegoes), or Skype (same username).

brainrape commented:
regarding file names:
```
src/
  Data/
    Json/
      JSemantic.purs
      JTable.purs
      JTable/
        Internal.purs
```
like this?

brainrape commented:
I've addressed most of the above comments, except:
- anything in `StrongCheck.purs`
- `Array.Unsafe.head` use for `JPath` (no other unsafe ops left)
- tuple rendering: please give me some examples to work with

Maximum width of homogeneus arrays treated as tuples is now an option.

brainrape commented:
I think I've addressed all your comments, except these issues:
- tuple rendering: please provide an example and the desired output
- additional examples: please provide some examples so I can include them

API changes: 
`th` now gets the current label as well as the path.
```purescript
TableStyle.th :: String -> JPath -> Markup
```
New option that controls the maximum size of homogeneous arrays treated as tuples (rendered horizontally).
```purescript
JTableOpts.maxHomoTupSize :: Number
```



Hope this will do.

jdegoes commented:
Hi @brainrape, sorry for the delay. I'll have a look at this today and provide lots of examples of tuples.

jdegoes commented on src/Data/Json/JSemantic.purs:
```diff
@@ -0,0 +1,130 @@
+
+parseX :: forall a. Regex -> (String -> Maybe a) -> (a -> JSemantic) -> String -> Maybe JSemantic
+parseX regexp parser constr s = 
+  match regexp s >>= tail >>= head >>= \s -> constr <$> parser s
```
FYI this can also be written as:

```purescript
match regexp s >>= tail >>= head >>= parser <#> constr
```

Although it would probably be more efficient to `case` on the regexp array and extract out the 2nd element rather than chaining bind through tail / head.

brainrape replied:
replaced with `case`

jdegoes commented on src/Data/Json/JTable.purs:
```diff
@@ -0,0 +1,72 @@
+import Text.Smolder.Markup (Markup(..), MarkupM(..), Attributable, attribute, (!), text)
+import Data.Foldable (mconcat)
+
+-- type JPath = [String]
```
Dead code.

brainrape replied:
actually, comment to show the type of JPath, since it is exported but not defined here

jdegoes commented on src/Data/Json/JTable/Internal.purs:
```diff
@@ -0,0 +1,218 @@
+toPrim :: Json -> Maybe JsonPrim
+toPrim = foldJsonP Just (const Nothing) (const Nothing)
+
+zipWithIndex xs = zip xs (0 .. ((length xs) - 1))
```
All top-level functions should have typos written out so they're easier to understand and appear in docs.

brainrape replied:
fixed

jdegoes commented on src/Data/Json/JTable/Internal.purs:
```diff
@@ -0,0 +1,218 @@
+
+zipWithIndex xs = zip xs (0 .. ((length xs) - 1))
+
+orelse f g x = case f x of Just v -> v 
```
Ditto. Also this looks like the `maybe` function defined in `Data.Maybe`.

brainrape replied:
I needed something like a lazy `<|>` since it was slowing down rendering.
Fixed: added type signature, refactored.I Left the last parens in the type to highlight the grouping.
```
orelse :: forall a b. (a -> Maybe b) -> (a -> b) -> (a -> b)
orelse f g = fromMaybe <$> g <*> f 
```

jdegoes commented on src/Data/Json/JTable/Internal.purs:
```diff
@@ -0,0 +1,218 @@
+
+-- maybe return the width of a tuple composed of primitive values
+widthOfPrimTuple :: Number -> JPath -> [Json] -> Maybe Number
+widthOfPrimTuple hS path ja =
```
I'll add some examples at the top-level PR comments section to clarify this. Currently it looks for homogeneity in the elements of a single tuple, not across the ith indices of vertical tuples.

brainrape replied:
The results from `widthOfPrimTuple` are later recursively merged using `tMergeArray` in `tFromJson`.

jdegoes commented on src/Data/Json/JTable/Internal.purs:
```diff
@@ -0,0 +1,218 @@
+type Table = [[Cell]]
+
+-- header data
+data Tree = T String JPath Number Number [Tree]
```
The width and height are implicitly contained in this data structure. Again, storing them only allows them to diverge and complicates the code. Also both a label and a complete `JPath` are stored, whereas a `JPath` can be constructed in any top-down traversal of the tree (so I'm not sure I see the point).

brainrape replied:
It is true, the path `p` can be calculated top-down, and `w` and `h` can be calculated bottom-up. 
However, these are all produced when the tree is first built up, never updated.
The tree is traversed twice more, both top-down:
- to render the header in `toRows`
- to produce a table structure for the data in `cFromJson`
These two traversals don't need to pass the extra top-down and bottom-up state and calculate the results, because the data is already available. Besides not duplicating work this also saves stack space and code complexity. However, it does use somewhat more heap space, but I don't think that is a concern since the number of nodes is equal to the number of labels in the json.

brainrape replied:
I thought of doing everything in one traversal but that is also difficult because the result is nested and processed differently.

Still, I might be on the wrong track here, so if you disagree, please illustrate what you propose with some code.

jdegoes replied:
For `JPath` elimination, my idea was to create a function similar to this:

```purescript
foldDownTree :: forall z. (Semigroup z) => (Tree -> JPath -> z) -> z
```

So your folding function has the current node, the path to the node (assembled using the labels in a top-down fashion), and can return anything as long as it can be merged with the final result (most likely `Table`??? Or some intermediate, mergeable structure?).

For the width and height, it's possible something simpler would suffice, which is just a recursive computation of the width and height (possibly memoized using `purescript-memoize` to avoid recomputation).

The reason I'm harping on this so much is because usually, these sorts of changes that eliminate duplication of information can have a *big* effect on the readability of code, its defect rate, and the ease with which it can be maintained over time in response to new feature enhancements. What are your thoughts on the above approach?

jdegoes commented on src/Data/Json/JTable/Internal.purs:
```diff
@@ -0,0 +1,218 @@
+-- add child to tree, unify if exists
+tMergeArray :: Tree -> Tree -> Tree
+tMergeArray (T l p w h k) nt@(T nl np nw nh nk) =
+  let i = findIndex (\n -> last np == last (n # tPath)) k in case k !! i of
```
This recomputes `last np` potentially many times. Not a big deal, just wanted to point it out.

brainrape replied:
fixed

jdegoes commented on src/Data/Json/JTable/Internal.purs:
```diff
@@ -0,0 +1,218 @@
+  , JPath(..), Table(..)
+  , tFromJson, tMergeArray, widthOfPrimTuple, padTree
+  , cFromJson, cMergeObj, mergeObjTuple
+  , foldJsonP, toPrim, zipWithIndex, orelse, strcmp, _nattr, _cspan, _rspan
```
Do all these functions need to be exported? Especially strcmp, _nattr, _cspan, _rspan, etc. But ideally as much of these can be hidden as possible.

brainrape replied:
I gather it's ok to export everything in  in `.Internal` modules, for testing and reuse.

jdegoes replied:
I'd prefer if we only exported the ones which are *actually called* from outside the module, because a lot of these are helper functions and could just as easily be defined in a `where` clause (well, except for repetition).

brainrape commented:
I've responded to the comments above. Fixed some stuff, backed the rest.

jdegoes commented:
OK, so I'm getting really close to merging this in. Before I do, we are going to provide the following:

1. Examples to demonstrate "vertically uniform" tuples and expected rendering, to clarify a bunch of questions around this issue.
2. Examples of JSON to add to the example base (if any of these have an issue in rendering, we will note them).

I'm going to work on (1) right now, and @cryogenian will begin work on (2), we will try to complete today but it may slip to tomorrow.

As far as revisions, I would *like* to see an approach that minimizes the amount of duplicated information (and I think you would also find it interesting), but I don't consider that a blocker to merging. At this point, I only have a couple blockers and I believe most of them are already addressed:

1. No uses of `Unsafe` in the library (I think we're there now, right?).
2. Using third-party libraries as much as possible in the examples (close or already there?).
3. Moving from "import all" style import declarations (e.g. `import Data.Array`) to selective imports (e.g. `import Data.Array(head, tail)`).

After we've added the examples, addressed the tuple rendering, and done a touch more cleanup, we'll be ready to merge this project in.

Thanks again for all your work on this, I am really excited about this library and think it will be tremendously useful to the broader Purescript community!

jdegoes commented:
I have added two more examples in a `Tuples` section: https://github.com/slamdata/purescript-jtable/blob/master/README.md

Happy to add more if you like, just let me know.

brainrape commented:
#### Blockers:
1. no `Unsafe` use left
2. Not yet, I Reduced the DOM operations to some simple FFI though. `purescript-simple-dom` 
3. Cleaned up imports, everything is explicit now.

#### Tuples

I've added the two new examples as `Example 2` and `Example 3` to `try.html`
Both new examples render correctly. `Example 2` renders as specified with default settings (sans css), and `Example 3` renders as specified when either the `Homogeneous Tuples Width 5` example renderer is selected, or any other datatype besides Number is included in one of the arrays (making it a heterogeneous array, which is always a tuple).

According to the original spec:
```
- Heterogeneous arrays of the same length are classified as tuples.
- Homogeneous arrays of differing lengths are classified as lists.
```
This is compatible with the implementation, and the unspecified cases (homogeneous arrays of equal lengths, heterogeneous arrays of differing lengths) support configuration through the `maxHomoTupSize` option. I've detailed this in my first comment, although the option is a newer development.


So, that leaves the DOM in the examples to fix.

I enjoyed working on this project a lot. I hope you'll excuse my noob nonsense, "it's just javascript" attitude, and gradual typing approach.

brainrape commented:
Hm, it turns out homogeneous arrays of differing lengths ar classified as tuples when they are below `maxHomoTupSize`.
For example, try this:
```
[{"a":[1,2]}, {"a":[5,6,7]}]
```
It looks kind of right though, that might be a "missing value".
So, is this ok, or should I fix this, or make it configurable?

jdegoes commented:
How about this behavior:

1. Primitive arrays are always rendered as horizontal tuples, regardless of length (requires null-padding to longest length unless there's some way to introduce heterogeneity using colspan magic or something).
2. Object arrays are always rendered as vertical lists (obviously with labels pulled out into the table header).
3. Array arrays are rendered as per their children (see above).

This is a very simple rule, easy to predict, and I think it will produce what users expect in most cases. Also we could delete the tuple cutoff configuration option. What do you think?

I'm glad you enjoyed the project and hope to see you continue engaging with the Purescript community. There's a lot of fun projects to work on and we'll probably do more bounty ones, too.

brainrape commented:
1. This is the case for heterogeneous arrays of primitives (that have more than one primitive type present).  The spec differentiates between homogeneous and heterogeneous tuples. Is this information irrelevant then?
Also, if in another branch of the json at the same place there are sub-labels, the whole column will be rendered vertically.
2. This is the case, except when the keys are disjunct, in which case the objects are collapsed into one (part of spec).
3. The arrays are ultimately flattened, but if the deepest level is a tuple, a vertical list of horizontal tuples is rendered.

brainrape commented:
I've looooong waited for a decent solution to the javascript problem in the form of a statically typed functional programming language. In the "almost" category, Roy brought hope and Elm showed how concise and elegant it could be, but Purescript has all the best features in one.
Funny, it doesn't seem so important that Purescript generates readable js since I very rarely have to look at it (except when I mess up FFI).

When I'm done with this, I'd like to write some games, and add some Elm-style libraries, and do some work on the compiler itself.
I'm also curious to see how native backends evolve. I'm pondering an Idris backend, since as far as I can tell Purescript is nearly a proper subset of Idris so it should be relatively easy to hook up codegen, and Idris in turn has several backends, including llvm. Then there's FFI and perhaps a libvu-based runtime for API-compatibility with node but compiled to native code. I'd push that.

cryogenian commented:
Hello! 
There are some examples of surprising behaviour. 

```json
{
  "objList" :[
 {"foo": 1,
  "bar": 2},
  {"foo": 2,
   "bar":3},
  {"foo": 3},
  {"bar": 12, 
   "foo": null}
  ],
  "fieldList": {
    "foo": [1, 2, 3, 4, 5],
    "bar": [1, 2, 3, 4, null]
  }
}
```
I believe <code>objList</code> and <code>fieldList</code> should be rendered same way.

```json

{
  "foo": {"bar": {"baz": {"quux": [1, 2, 3, 4, 5]}}},
  "bar": {"baz": {"quux": {"foo": [1, 3, 4, 5, 5]}}},
  "baz": [1, 2, 3, 4, 5],
  "quux": [{"o": 1}, {"o": 2}, {"o": 3}, {"o": [1, 2, 3, 4]}, {"o": 5}]
}
```
Don't fourth row in fields ```foo```, ```bar``` and ```baz``` must be spaned? 


Atomic values can be rendered. Not sure it's very important bug, but ```null``` is not valid json. 


```json
{
  "foo": [1, 2, 3, 4, 5],
  "bar": [
     {"baz": [1, 2, 3, 4, 5]},
     {"baz": [1, 2, 3, 4, 5]}
   ]
}
```
Surprising behaviour since I do not suppose that ```bar.baz``` is flattened list. I don't even think that they can be directly mapped to ```foo```. One real sample of ```foo``` is common metadata and ```bar.baz``` item statistics. 



```json 
  {"foo": []}
```
It renders two rows. 


```json
{
  "foo": [],
  "bar": [],
  "baz": [null],
  "quux": null,
  "fooo": [null, [], null]
}
```
```json
{
  "foo": [],
  "bar": [],
  "baz": [null],
  "quux": null,
  "fooo": [null, null, null]
}
```
This jsons produce different tables. And ```[null, [], null]``` is rendered vertically but ```[null, null, null]``` is rendered horizontally. 


```json
[
  {"foo": "bar"},
  {"bar": "baz"},
  {"baz": "quux"}
]
```
```json
[
  {"foo": "bar"},
  {"bar": "baz"},
  {"baz": "quux"}, []
]
```
```json
[
  {"foo": "bar"},
  {"bar": "baz"},
  {"baz": "quux"}, {}
]
```
```json
[
  {"foo": "bar"},
  {"bar": "baz"},
  {"baz": "quux"}, null
]
```
This three json have different and very interesting output, hope them help to figure out  what to do with ```[]``` and ```null```. 

That's all that I've found, though.

brainrape commented:
Thanks for the examples. Here's a detailed response with numbering according to code blocks.

#### 1. 

If you mean both `foo` and `bar` columns from both `objList` and `fieldList` should be rendered vertically, note that fieldList.bar[4] is of a different type (`null`) than the rest of the array, and so this heterogeneous array is typed a tuple and rendered horizontally, so this is the expected behavior, as per the specification. However, please try the renderer `Homogeneous Tuples Width 5` and see if that is what you expected.


#### 2.

Since `quux` has a different nesting structure (list of objects), it cannot be determined that elements should be aligned with the lists in the other labels. A list of `quux.o` is probably not a list of `baz`, so their values and nested arrays are not aligned.

Also, null is in fact a valid JSON primitive value.

#### 3.

Related to the previous case.

Since arrays can be nested arbitrarily, they are flattened into parallel 1 cell high items. Alignment is kept at object level. Precisely because `foo` can't be mapped to `bar.baz`, the lists are simply rendered in parralel, vertically, without keeping the outer array aligned.

Please,
* Try the renderer `Homogeneous Tuples WIdth 5` and see if that is what you expected.
* Try the following json, and let me know if that is how you think it should look

```
[{
  "foo": 1,
  "bar": {"baz": [1, 2, 3, 4, 5]}
},{
  "foo": 2,
  "bar": {"baz": [1, 2, 3, 4, 5]}
},{
  "foo": 3
},{
  "foo": 4
},{
  "foo": 5
}]
```


#### 4.

Fixed.


#### 5. 6.

Both of these render as intended. `[null, [], null]` contains non-primitive values so it's a list (vertical). Since the `[]` in the middle doesn't contain any values, it doesn't produce any cells.
`[null, null, null]` is an array of primitives, which is homogeneous but below `maxHomoTupSize` so a tuple (horizontal).


#### Last 4

The last 4 code blocks work as intended:
1 and 3 are equivalent: they are tuples of objects with disjunct keys, rendered on one row.
2 and 4 are similar in that the array is treated as a list, not a tuple; but in 2 the `[]` doesn't contain any primitive values so doesn't produce any cells, and in 4 the `null` primitive value produces a cell.

cryogenian commented:
If only json have ```undefined``` I'd never ask about first question. I think that sparsed arrays can be emulated only by using ```null``` as unexistent value. So, yes I can indeed show it by using another structure, but what if I have ```[1, null, 2]```? 

If we can't say that ```quux.o``` can be mapped to ```baz``` why they aligned? It's easier to see that there is no relationship if one column consist of 8 rows and other of 5. And if we have exactly 5 ```quux``` why don't we can suggest such relationship between ```quux``` and ```baz```. 

Though we can't say that there is relationship, but we see it :) because when I see that two values are in one row I suggest they somehow connected. Why don't just split ```foo``` column to 5 rows and ```bar.baz``` to ten?

brainrape commented:
I ported most of the examples to `purescript-jquery` but they still need 2 FFI'd functions.

brainrape commented:
@cryogenian I'm not sure what you mean. Please take a look at the Olympic Games sample json. The `sports` column on the left and `events` on the right are separate from and parallel to the `/olympics/olympic_games/medals_awarded` group of columns. I think this is a good way to render these lists. Do you disagree? How else do you suggest rendering parallel lists?

brainrape commented:
Let's take a small example.
```
{
  "foo": ["one", "two", "three", "four", "five"],
  "bar": [1, 2, [3, 3], 4, 5]
}
```
All the "foo" cells are one row tall. Do you suggest that `three` should be two rows tall?
If yes, I disagree since generally these are columns of different things, and aligning them could be confusing. However, if the nesting structure is the same (perhaps because the columns are related), the data will line up anyway.

Please note, if all of the elements in a list are tuples, it will be rendered as a list of tuples, for example:
```
{
  "foo": ["one", "two", "three", "four", "five"],
  "bar": [[1, 0, 0], [2, 0, 0], [3, 0, 0], [4, 0, 0], [5, 0, 0]]
}
```
I think this is an adequate way of treating unknown data, with consistent results.
True, some information is lost since arrays are flattened, but I'm not sure how else a table could be consistently rendered.  
If I still didn't get what you mean, a minimal test case with expected resulting table could help illustrate the problem.

cryogenian commented:
One can render parallel lists by choose lesser common multiply as number of rows and then provide for every cell rowspan. I believe it is possible. I.e. you have 
```json
{
  "foo": [1, 2],
  "bar": [1, 2, 3]
```
And this can be rendered to 
______
|     |  1  |
|  1 |___|
|__ |  2  |
|     | _ _|
|  2 |  3  |
| __|_ _ | 
all numbers should be vertical aligned in middle

I suggest this approach is very clear show that there is no relations between ```foo``` and ```bar```. Yes, they will be shown as ```foo``` related with two ```bar``` if 
```json
{
  "foo": [1, 2],
  "bar": [1, 2, 3, 4]
}
```
But it seems more clear than ```foo``` having two empty cells. 
Moreover if we can use ```null``` as unexistent value it's possible to render hidden but very possible relations like
```json
{
  "foo": [1, 2, 3],
  "bar": [1, 2, 3, null, 5, 6]
}
```

cryogenian commented:
Oh, my example is not very informative it should be something like this 
```html
<table>
<tr id="1">
  <td rowspan="3"> 1
  </td>
  <td rowspan="2"> 1
  </td>
</tr>
<tr id="2">
</tr>
<tr id="3">
  <td rowspan="2">2</td>

</tr> 
<tr id="4">
  <td rowspan="3">2</td>
</tr>
 <tr id="5">
  <td rowspan="2">3</td>
  </tr>
 <tr id="6">
 </tr>
</table>
```

brainrape commented:
Hm, are you suggesting to vertically justify arrays of different length? Ie in the example above, the total number of rows should be 6, the least common multiplier of 2 and 3? I thought about this as well, but instead decided for padding the shorter array(s) to the length of the longest array, since that will produce a simpler result, and prevents an explosion of rows which would make (browser) rendering slow for even medium sized datasets. For example, two arrays, 20 and 21 in length, would produce 410 table rows. Add in another parallel array, of possibly coprime length, and you see where this is going. It adds O(n^a) space/time complexity (for the browser layout engine) where n is the maximum length of arrays and a is the number of parallel arrays.
Could be an option though. I don't see the need for it, but I'll add it if you prefer.

As for the last json example with `null` in the middle, I'm not sure what result you expect, but the array is rendered as a tuple (horizontally) since it is heterogeneous.

jdegoes commented:
> This is the case for heterogeneous arrays of primitives (that have more than one primitive type present). The spec differentiates between homogeneous and heterogeneous tuples. Is this information irrelevant then? 

I think the answer should be "yes". In all the cases I have looked at, the most salient point is whether the array consists of array primitives or whether it contains objects. If it consists of array primitives (or arrays of primitives, etc.), then a horizontal, nameless, "tuple"-rendering efficiently uses the available space, and results in a compact, relatively easy rendering. I also think when users intend tuples (e.g. lat/long, or geo polygon), they generally use primitives anyway.

If it consists of an array of objects, then the labels are already pulled into the header, which organizes all the array elements that have the same structure, and the vertical encoding quite nicely encodes the idea that there are "many" of these things per row in the dataset.

So I think rather than having cutoffs for tuple lengths, and rather than looking for homogeneous or heterogeneous arrays, the simplest and most visually consistent thing is to see if an array consists of all primitives or not. If all primitives, use the tuple / horizontal rendering, otherwise, use the vertical rendering. This logic is simpler, more uniform (no need for an arbitrary "cutoff"), and seems to produce good results in all the test examples, as well as various other cases I can demonstrate.

After we get this thing merged, I'll create some tickets that we'll probably end up working to better handle nested arrays (i.e. your original idea of being able to infer the structure of the JSON from the way it's rendered, which currently cannot be done for nested arrays), and figure out how that works with the above. Arrays are definitely the trickiest part about this.

I haven't delved deep into the above discussion between you and @cryogenian, and will do so at some point, but in the meantime, please add those examples to the test application (e.g. "Malformed Example #1", etc.).

brainrape commented:
Ok, I'll update things so heterogeneity is not checked, only that all the items are primitives.
I think the cutoff is necessary though, since you probably don't want to render 100 items horizontally, and they could even be long strings.
So the new option would be `maxTupleSize`, with default say 10. Agreeable?

jdegoes commented:
@brainrape Good idea, sounds good to me.

brainrape commented:
Added examples, updated behavior as discussed.
All arrays of primitives (upto length maxTupleSize) are treated as tuples and rendered horizontally if possible.

brainrape commented:
We should be good to merge. Or is there anything else?

jdegoes commented:
I will test the above changes on Monday, and hopefully get this thing merged!

jdegoes commented:
Hi @brainrape, there are some changes I'd like to make, but I think you've already earned the bounty so I'll go ahead and merge this in. Thanks for your work bringing this to the PureScript community! Let me know your email and I'll have Jeff reach out.

brainrape commented:
Great! I'm looking forward to the issues. martonboros at gmail

JeffACarr commented:
Hi Martin,

I'll pay the bounty in the next few days.  I pay using Paypal, so please
confirm your account limits will accept this size transfer ($2000 USD).
Also, if you can create a Paypal invoice and send it to my account jeff at
slamdata dot com.

Once confirmed I'll make payment.

Thanks for you work on the jtable project.

Jeff

On Tue, Mar 17, 2015 at 8:11 PM, Márton Boros <notifications@github.com>
wrote:

> Great! I'm looking forward to the issues. martonboros at gmail
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/slamdata/purescript-jtable/pull/11#issuecomment-82684258>
> .
>



-- 
*Jeff A. Carr*
*Co-Founder & CEO*
*303-717-2091*
<http://slamdata.com/>
